samples/bluetooth/alexa_gadget/src/gadgets/gadgets_profile.c

```diff
@@ -0,0 +1,1231 @@
+	const uint8_t *data, uint16_t len)
+{
+#if CONFIG_GADGETS_CAPABILITY_ALERTS_ENABLE
+	bool success;
+	pb_istream_t stream_in;
+
+	stream_in = pb_istream_from_buffer(data, len);
+
+	if (strcmp(name, "SetAlert") == 0) {
+		success = pb_decode(
+					&stream_in,
+					&alerts_SetAlertDirectiveProto_msg,
+					&protobuf_decoded.set_alert);
+		if (!success) {
+			LOG_ERR("SetAlert decode error");
+			return -EINVAL;
+		}
+
+		struct gadgets_evt evt = {
+			.type = GADGETS_EVT_SETALERT,
+			.parameters.set_alert = NULL,
+		};
+
+		if (protobuf_decoded.set_alert.has_directive &&
+			protobuf_decoded.set_alert.directive.has_payload) {
+			evt.parameters.set_alert =
+				&protobuf_decoded.set_alert.directive.payload;
+		}
+
+		callback(&evt);
+		return 0;
+	} else if (strcmp(name, "DeleteAlert") == 0) {
+		success = pb_decode(
+					&stream_in,
+					&alerts_DeleteAlertDirectiveProto_msg,
+					&protobuf_decoded.delete_alert);
+		if (!success) {
+			LOG_ERR("DeleteAlert decode error");
+			return -EINVAL;
+		}
+
+		struct gadgets_evt evt = {
+			.type = GADGETS_EVT_DELETEALERT,
+			.parameters.delete_alert = NULL,
+		};
+
+		if (protobuf_decoded.delete_alert.has_directive &&
+			protobuf_decoded.delete_alert.directive.has_payload) {
+			evt.parameters.delete_alert =
+				&protobuf_decoded.delete_alert.directive.payload;
+		}
+
+		callback(&evt);
+		return 0;
+	}
+#endif
+	return -ENOTSUP;
+}
+
+static int directive_handler_custom(
+	struct bt_conn *conn, const char *name,
+	const uint8_t *data, uint16_t len)
+{
+#if CONFIG_GADGETS_CAPABILITY_CUSTOM_ENABLE
+	if (!protobuf_decoded.directive_parser.has_directive) {
+		LOG_DBG("Empty custom directive");
+		return 0;
+	}
+
+	struct gadgets_evt evt = {
+		.type = GADGETS_EVT_CUSTOM,
+		.parameters.custom_directive.name = name,
+		.parameters.custom_directive.payload =
+			protobuf_decoded.directive_parser.directive.payload.bytes,
+		.parameters.custom_directive.size =
+			protobuf_decoded.directive_parser.directive.payload.size,
+	};
+
+	callback(&evt);
+	return 0;
+#endif
+	return -ENOTSUP;
+}
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	if (err) {
+		LOG_WRN("Connection error: %d", err);
+		return;
+	}
+
+	conn_state.conn = bt_conn_ref(conn);
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	if (conn_state.conn) {
+		bt_conn_unref(conn_state.conn);
+	}
+
+	memset(&conn_state, 0, sizeof(conn_state));
+
+	encoded_buffer_release();
+}
+
+static void security_changed(struct bt_conn *conn, bt_security_t level,
+			     enum bt_security_err err)
+{
+	int mtu_err;
+
+	if (err) {
+		LOG_WRN("Security failed: level %u err %d", level, err);
+		return;
+	}
+
+	if (level < BT_SECURITY_L2) {
+		LOG_WRN("Insuficcient sec level");
+		return;
+	}
+
+	conn_state.encrypted = 1;
+
+	mtu_err = bt_gatt_exchange_mtu(conn_state.conn, &exchange_params);
+	if (mtu_err) {
+		LOG_WRN("bt_gatt_exchange_mtu: %d", err);
+	}
+}
+
+static void pairing_confirm(struct bt_conn *conn)
```

Comment:
I wonder if it would be better to move it to `main` file and supports also passkey or numeric comparison ?

Reply:
Yes, good idea regarding MITM. I will test this with the Alexa Echo. The reason for placing the function within the profile is that it is required for the Profile to function properly. Otherwise a developer has to fully review main.c to understand which parts might be relevant for the Profile

Reply:
The Echo I tested with did not utilize MITM.
I'd like to keep the pairing_confirm function within the profile for now

Reply:
Ok